### PR TITLE
Behavioral Cloning

### PR DESCRIPTION
**Pull request by krzentner**
Fixes / Implements #380

### CODE REVIEW

File: src/garage/torch/algos/bc.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    """Behavioral Cloning.
+
+    Based on Model-Free Imitation Learning with Policy Optimization:
+        https://arxiv.org/abs/1605.08478
+
+    Args:
+        env_spec (garage.envs.EnvSpec): Specification of environment.
+        learner (garage.torch.Policy): Policy to train.
+        batch_size (int): Size of optimization batch.
+        source (garage.Policy or Generator[garage.TimeStepBatch]): Expert to
```
Comment (ryanjulian) [inline src/garage/torch/algos/bc.py:27]: out-of-scope for this PR: what do you think about defining a Dataset API, which would unify things like this (in addition to accessing the replay buffer, or on-policy sampling).

basically `TimeStepDataset = Generator[garage.TimeStepBatch]` and `TrajectoryDataset = Generator[garage.TrajectoryBatch]`

alternatively, do this but use the torch and tf first-party APIs

**Comment by ryanjulian** [comment]
You LOVE to see it!

File: src/garage/torch/algos/bc.py (hunk 1 above)
Comment (krzentner) [reply src/garage/torch/algos/bc.py:27]: Yeah, I'm not sure if the first party APIs are best to use or not at this point. It seems unlikely we'd go wrong using a more restrictive API for the mean time.

File: examples/torch/bc_point.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        Args:
+            observation (np.ndarray): Observation from the environment.
+
+        Returns:
+            tuple:
+                * np.ndarray: Action.
+                * dict[str, np.ndarray]: Agent info (empty).
+
+        """
+        return self.goal - observation[:2], {}
```
Comment (ahtsan) [inline examples/torch/bc_point.py:40]: Does this assume observation's dimension is [batch, time, *dims]? I think we should mention it here.

File: examples/torch/bc_point.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    def __init__(self, env_spec, goal):
+        super().__init__(env_spec, 'OptimalPolicy')
+        self.goal = goal
+
+    def get_action(self, observation):
+        """Get action given observation.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
```
Comment (avnishn) [inline examples/torch/bc_point.py:32]: can you document this shape please

File: examples/torch/bc_point.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                * dict[str, np.ndarray]: Agent info (empty).
+
+        """
+        return self.goal - observation[:2], {}
+
+    def get_actions(self, observations):
+        """Get actions given observations.
+
+        Args:
+            observations (np.ndarray): Observations from the environment.
```
Comment (avnishn) [inline examples/torch/bc_point.py:46]: can you document this shape?

Comment (avnishn) [reply examples/torch/bc_point.py:46]: I think the variables to be used to describe the shape are outlined in a table under garage/_dtypes.py

File: examples/torch/bc_point_deterministic_policy.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    def __init__(self, env_spec, goal):
+        super().__init__(env_spec, 'OptimalPolicy')
+        self.goal = goal
+
+    def get_action(self, observation):
+        """Get action given observation.
+
+        Args:
+            observation (np.ndarray): Observation from the environment.
```
Comment (avnishn) [inline examples/torch/bc_point_deterministic_policy.py:31]: please document shapes. Variables to describe the shape can be found in a table under garage/_dtypes.py.

File: examples/torch/bc_point_deterministic_policy.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        Returns:
+            tuple:
+                * np.ndarray: Action.
+                * dict[str, np.ndarray]: Agent info (empty).
+
+        """
+        return self.goal - observation[:2], {}
+
+    def get_actions(self, observations):
```
Comment (avnishn) [inline examples/torch/bc_point_deterministic_policy.py:41]: please document shapes. Variables to describe the shape can be found in a table under garage/_dtypes.py

File: src/garage/_dtypes.py [hunk 7]
```diff
@@ -919,3 +921,28 @@ def from_time_step_list(cls, env_spec, ts_samples):
         ]
 
         return TimeStepBatch.concatenate(*ts_batches)
+
+    @classmethod
+    def from_trajectory_batch(cls, batch):
+        """Construct a TimeStepBatch from a TrajectoryBatch.
+
+        Args:
+            batch (TrajectoryBatch): TrajectoryBatch to convert.
```
Comment (avnishn) [inline src/garage/_dtypes.py:930]: please document shapes

File: examples/torch/bc_point.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+import numpy as np
+
+from garage import wrap_experiment
+from garage.envs import GarageEnv, PointEnv
+from garage.experiment import LocalRunner
+from garage.torch.algos import BC
+from garage.torch.policies import GaussianMLPPolicy, Policy
+
+
+class OptimalPolicy(Policy):
```
[the remark this answers is not in the document]
Comment (ryanjulian) [reply examples/torch/bc_point.py:12]: examples is not on the python path (all files are self-contained scripts), so no doesn't work. also having a library of utilities within examples is just strange. if it's so useful, why isn't it in garage?

i think this solution is fine.

File: examples/torch/bc_point.py (hunk 4 above)
Comment (ryanjulian) [reply examples/torch/bc_point.py:46]: alternatively, the entire body of this docstring can just refer to the policy interface.

File: src/garage/_dtypes.py (hunk 7 above)
Comment (ryanjulian) [reply src/garage/_dtypes.py:930]: what shapes? the dtype is well-defined. that's why we defined it! to stop typing shapes all the time.

**Pull request merged**